app/routes/api/verify.ts

```diff
@@ -0,0 +1,47 @@
+import type { ActionFunction, DataFunctionArgs } from "@remix-run/server-runtime";
+import { json } from "@remix-run/server-runtime";
+import { unprocessableEntity } from "remix-utils";
+import { SiweMessage } from "siwe";
+import { createUserSession, getNonce } from "~/services/session.server";
+import { createUser, findUserByAddress } from "~/services/user.server";
+import chainalysisAbi from "~/abi/chainalysis.json";
+import Web3 from "web3";
+import type { AbiItem } from "web3-utils";
+
+const RPC_URL = "https://mainnet.infura.io/v3/54fcc811bac44f99b84a04a4a3e2f998";
+
+export const action: ActionFunction = async (data: DataFunctionArgs) => {
+  const { message, signature } = await data.request.json();
+  const siweMessage = new SiweMessage(message);
+  const fields = await siweMessage.validate(signature);
+
+  if (await isAddressSanctioned(message.address)) {
+    return unprocessableEntity({ message: "Sanctioned Address" });
+  }
+  const nonce = await getNonce(data.request);
+  if (fields.nonce !== nonce) {
+    return unprocessableEntity({ message: "Invalid nonce" });
+  }
+
+  const userAddress = message.address;
+  let user = await findUserByAddress(userAddress);
+  if (!user) {
+    user = await createUser(userAddress);
+  }
+
+  await createUserSession({
+    request: data.request,
+    userId: user.id,
+  });
+
+  return json({ ok: true });
```

Comment:
Huh, are we sure this works? I think the action needs to actually return the response with the session.